cpp/src/arrow/util/bit_block_counter.h

```diff
@@ -87,6 +87,147 @@ struct BitBlockOrNot<bool> {
   static bool Call(bool left, bool right) { return left || !right; }
 };
 
+// Three Arguments
+template <typename T>
+struct BitBlockAndAnd {
+  static T Call(T left, T mid, T right) { return left & mid & right; }
+};
+
+template <>
+struct BitBlockAndAnd<bool> {
+  static bool Call(bool left, bool mid, bool right) { return left && mid && right; }
```

Comment:
There is no need to specialize all these bitwise operations for `bool` as the language guarantees that for boolean types: `(bool)(a & b) == a && b`. Note the explicit cast which corresponds to the return type `T` in these functions. This simplification also applies to the existing binary bitwise operations, so we can resolve this in a follow-up PR.

Reply:
Done

Reply:
Note, the only bool specialization needed would be for negating bool variables. I submitted [this PR with these changes](https://github.com/apache/arrow/pull/12052) that if they get accepted you can make use of the added`BitNot()` to wrap negations.

Reply:
Reverted this change until https://github.com/apache/arrow/pull/12052 is resolved.

Reply:
Actually, the only combination used in this PR is `BitBlockAndAnd`, so you can remove all the other ones. There is no reason to have dead code. Also, ARROW-15220 got merged so you can see the new form for these operators. Note: It will change [their use as template parameters](https://github.com/apache/arrow/blob/master/cpp/src/arrow/util/bit_block_counter.h#L287-L288) and how [`Call` is invoked](https://github.com/apache/arrow/blob/master/cpp/src/arrow/util/bit_block_counter.h#L322).

Reply:
Done.